src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala

```diff
@@ -112,6 +123,56 @@ class RegularSync(
       }
   }
 
+  /**
+    * Handles NewHashesMessege, should only cover this message when we are top of the chain
+    */
+  def handleNewBlockHashesMessages: Receive = {
+    case MessageFromPeer(NewBlockHashes(hashes), peerId) =>
+      val maybePeer = peersToDownloadFrom.find(peer => peer._1.id == peerId)
+      //we allow asking for new hashes when we are not syncing and we can download from specified peer and we are
+      //top of the chain
+      if (notDownloading() && topOfTheChain && maybePeer.isDefined) {
```

Comment:
As they are similar, should we maybe also check that we are in the `topOfTheChain` when handling `NewBlock` messages?

In relation to that, could it be the case that our client stops being in the top of the chain (as if there is any network issues)? Could not detecting that affect properly handling `NewBlockHashes` messages?

Reply:
Yep we could definitely check it also for `NewBlock` message, as currently even if we are still far away from a top, we are receiving new blocks and try to import them. It ends up  in https://github.com/input-output-hk/mantis/blob/e6d14f5abbae5d1ad3257df39d8793a8c0997f67/src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala#L105
So it would be minor optimisation.

As for possible network issues, currently, i decided to go for most simple solution i.e once we reach top of the chain once, we are turning on handling `NewBlockHashes` permanently. It can have some downsides like, if we will be left behind by to many blocks because of this network issue, we will be receiving too fresh `NewBlockHashes` and handling them will slow down our regular `askForHeaders` cycle. (though I am still not sure how much, it needs to be tested in action)